### PR TITLE
Refactor authentication to be available for Google accounts.

### PR DESCRIPTION
**Pull request by ViliMakinen**
Refactor authentication to be available for Google accounts.

Facebook authentication is not in place yet.

### CODE REVIEW

File: infra/modules/app/ecs.tf [hunk 1]
```diff
@@ -46,13 +51,53 @@ resource "aws_ecs_task_definition" "cheerful" {
       {
         "name": "DATABASE_URL",
         "value": "postgresql://${aws_db_instance.app.username}:${aws_db_instance.app.password}@${aws_db_instance.app.address}:${aws_db_instance.app.port}/${aws_db_instance.app.db_name}?schema=public"
+      },
+      {
+        "name": "JWT_SIGNING_KEY",
+        "value": "${random_password.jwt_signing_key.result}"
+      },
+      {
+        "name": "CLIENT_ID",
```
Comment (RoopeHakulinen) [inline infra/modules/app/ecs.tf:60]: **chore**: remove

File: frontend/src/app/side-menu/side-menu.component.ts [hunk 2]
```diff
@@ -7,15 +7,16 @@ import { User } from '../user';
...
-  constructor(
-    public menuService: MenuService, 
-    public authService: AuthService
-    ) {
+  constructor(public menuService: MenuService, public authService: AuthService) {}
 
-    }
+  logOut(): void {
+    this.authService.signOut().subscribe(() => {
+      this.authService.navigateToApp();
```
Comment (RoopeHakulinen) [inline frontend/src/app/side-menu/side-menu.component.ts:19]: **question:** navigate to root, not /app, right?

File: frontend/src/app/home/home.component.ts [hunk 3]
```diff
@@ -4,10 +4,12 @@ import { AuthService } from '../auth.service';
...
   templateUrl: './home.component.html',
-  styleUrls: ['./home.component.scss']
+  styleUrls: ['./home.component.scss'],
 })
 export class HomeComponent {
-
   constructor(public authService: AuthService) {}
 
+  logIn(): void {
+    window.location.replace('/api/oauth2/google?redirect_url=/app');
```
Comment (RoopeHakulinen) [inline frontend/src/app/home/home.component.ts:13]: **issue:** the `redirect_url` should be the current path over static `/app`

File: frontend/src/app/auth.service.ts [hunk 4]
```diff
@@ -1,58 +1,26 @@
...
     this.router.navigate(['/app']);
   }
 
-  isSignedIn(): Observable<boolean> {
-    return this.authService.authState.pipe(
-      map((user) => user !== null),
-      take(1),
-    );
+  signOut(): Observable<boolean> {
+    return this.http.get<boolean>('/api/users/logout');
```
Comment (RoopeHakulinen) [inline frontend/src/app/auth.service.ts:20]: **suggestion:** use `unknown` or `void` as type as there shouldn't really be anything returned by the backend

File: backend/src/users/users.controller.ts [hunk 5]
```diff
@@ -12,6 +12,17 @@ export class UsersController {
...
 
+  @Get('me')
+  async getMe(@Req() req: any): Promise<User> {
+    return this.usersService.findOne(req.user.id);
+  }
+
+  @Get('logout')
+  async logout(@Res() res: any): Promise<void> {
+    res.clearCookie('AuthToken', { secure: true, httpOnly: true });
+    res.redirect('/');
```
Comment (RoopeHakulinen) [inline backend/src/users/users.controller.ts:23]: **question:** does this need to be a redirect? Should be possible to delete the cookies as part of an empty standard response as this is invoked as an API call

File: frontend/package-lock.json [hunk 6]
```diff
@@ -8,7 +8,6 @@
       "name": "cheerful",
       "version": "0.0.0",
       "dependencies": {
-        "@abacritt/angularx-social-login": "1.2.2",
```
Comment (RoopeHakulinen) [inline frontend/package-lock.json:11]: **question:** did you forget to commit package.json?

File: backend/src/auth/oauth2-google-strategy.service.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+export class OAuth2GoogleStrategy extends PassportStrategy(GoogleStrategy, 'oauth2-google') {
+  constructor(private authService: AuthService, private configService: ConfigService) {
+    super(
+      {
+        clientID: configService.get('oauth2-google.clientId'),
+        clientSecret: configService.get('oauth2-google.clientSecret'),
+        callbackURL: configService.get('oauth2-google.callbackUrl'),
+        passReqToCallback: true,
+        scope: ['profile', 'email'],
+        state: true,
```
Comment (RoopeHakulinen) [inline backend/src/auth/oauth2-google-strategy.service.ts:19]: **question:** how'd you make it work with state?

File: backend/src/auth/jwt-auth.guard.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { ExecutionContext, Injectable } from '@nestjs/common';
+import { AuthGuard } from '@nestjs/passport';
+import { Observable } from 'rxjs';
+
+@Injectable()
+export class JwtAuthGuard extends AuthGuard('jwt') {
+  canActivate(context: ExecutionContext): boolean | Promise<boolean> | Observable<boolean> {
+    const whitelistedPaths = ['/api/oauth2', '/api/logout'];
```
Comment (RoopeHakulinen) [inline backend/src/auth/jwt-auth.guard.ts:8]: **chore:** I don't think we need to whitelist the logout path as we don't redirect to it

File: backend/src/auth/auth.module.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+@Module({
+  imports: [
+    PassportModule.register({ defaultStrategy: 'jwt' }),
+    ConfigModule,
+    JwtModule.registerAsync({
+      imports: [ConfigModule],
+      useFactory: async (configService: ConfigService) => ({
+        secret: configService.get<string>('JWT_SIGNING_KEY'),
+        signOptions: { expiresIn: '48h' },
```
Comment (RoopeHakulinen) [inline backend/src/auth/auth.module.ts:19]: **issue:** we don't want to make users sign back in every two days so let's remove the option to not add expiration to the tokens

File: infra/modules/app/ecs.tf [hunk 10]
```diff
@@ -46,13 +51,53 @@ resource "aws_ecs_task_definition" "cheerful" {
       {
         "name": "DATABASE_URL",
         "value": "postgresql://${aws_db_instance.app.username}:${aws_db_instance.app.password}@${aws_db_instance.app.address}:${aws_db_instance.app.port}/${aws_db_instance.app.db_name}?schema=public"
+      },
+      {
+        "name": "JWT_SIGNING_KEY",
```
Comment (RoopeHakulinen) [inline infra/modules/app/ecs.tf:56]: **issue:** missing DOMAIN here

**Review by RoopeHakulinen — APPROVED** [review]

File: frontend/src/app/home/home.component.ts (hunk 3 above)
Comment (ViliMakinen) [reply frontend/src/app/home/home.component.ts:13]: Won't the current path always be '/', since this method is only called from the home page?

Comment (RoopeHakulinen) [reply frontend/src/app/home/home.component.ts:13]: True dat

File: infra/modules/app/ecs.tf (hunk 10 above)
Comment (ViliMakinen) [reply infra/modules/app/ecs.tf:56]:
```
     {
        "name": "DOMAIN",
        "value": "dev.getcheerful.com"
      }
```

Can the DOMAIN be just hardcoded like this?

File: backend/.env [hunk 11]
```diff
@@ -4,4 +4,8 @@
 # Prisma supports the native connection string format for PostgreSQL, MySQL, SQLite, SQL Server, MongoDB and CockroachDB.
 # See the documentation for all the connection string options: https://pris.ly/d/connection-strings
 
-DATABASE_URL="postgresql://postgres:sMgKQ3Wtj9pKZ9wdfh9Ufbd8@localhost:5432/postgres?schema=public"
\ No newline at end of file
+DOMAIN=http://localhost:4200
+DATABASE_URL=postgresql://postgres:sMgKQ3Wtj9pKZ9wdfh9Ufbd8@localhost:5432/postgres?schema=public
+CLIENT_ID=643585029299-4p9vpkh0tiscrfs1tpvhf8ok0q1v1t60.apps.googleusercontent.com
+CLIENT_SECRET=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
```
[the remark this answers is not in the document]
Comment (ViliMakinen) [reply backend/.env]: For the `CLIENT_SECRET`?

File: infra/modules/app/ecs.tf (hunk 10 above)
Comment (RoopeHakulinen) [reply infra/modules/app/ecs.tf:56]: It is available as `local.environment_domain`

File: backend/.env (hunk 11 above)
[the remark this answers is not in the document]
Comment (RoopeHakulinen) [reply backend/.env]: Fine, wrote this yesterday before we chatted

File: infra/modules/app/ecs.tf (hunk 10 above)
Comment (ViliMakinen) [reply infra/modules/app/ecs.tf:56]: Cool!

File: backend/src/auth/oauth2-google-strategy.service.ts (hunk 7 above)
Comment (ViliMakinen) [reply backend/src/auth/oauth2-google-strategy.service.ts:19]: In the end I didn't - tested it now one more time and using `state: true` doesn't work. I changed it back to `state: false`.

**Pull request merged**